### PR TITLE
SLING-9834 - [Sling Models] Caching bug with reused Servlet requests

### PR DESCRIPTION
**Pull request by cjelger**
- use ServletRequest attributes to cache sling models when the adaptable is a ServletRequest

### CODE REVIEW

**Comment by cjelger** [comment]
@justinedelson @cziegeler @raducotescu Can you please review the PR? Thanks.

**Review by cziegeler — COMMENTED** [review]
Lgtm in general, thanks.
If getOrCreateCache can be called concurrently for the same key, you might need to synchronize the creation.
For per request caching, you don't need a synchronized map nor SoftReference as everything is cleaned up at the end of the request - but that's probably a larger change.

File: src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java [hunk 1]
```diff
@@ -404,11 +430,7 @@ public boolean isModelClass(@NotNull Class<?> type) {
                         ModelType model = (ModelType) Proxy.newProxyInstance(modelClass.getType().getClassLoader(), new Class<?>[] { modelClass.getType() }, handlerResult.getValue());
 
                         if (modelAnnotation.cache()) {
-                            Map<Class<?>, SoftReference<Object>> adaptableCache = adapterCache.get(cacheKey);
-                            if (adaptableCache == null) {
-                                adaptableCache = Collections.synchronizedMap(new WeakHashMap<Class<?>, SoftReference<Object>>());
-                                adapterCache.put(cacheKey, adaptableCache);
-                            }
+                            Map<Class<?>, SoftReference<Object>> adaptableCache = getOrCreateCache(adaptable);
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:433]: Don't we already know here that there was no cache created for this `adaptable`? If a cache was available, we would have returned the result already - see https://github.com/apache/sling-org-apache-sling-models-impl/pull/21/files#diff-0c273224a9fa6e263f4381c5d4b8145aa2c6c89e75e0b62025eea5efa17edec9R401.

File: src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java [hunk 2]
```diff
@@ -421,11 +443,7 @@ public boolean isModelClass(@NotNull Class<?> type) {
                         result = createObject(adaptable, modelClass);
 
                         if (result.wasSuccessful() && modelAnnotation.cache()) {
-                            Map<Class<?>, SoftReference<Object>> adaptableCache = adapterCache.get(cacheKey);
-                            if (adaptableCache == null) {
-                                adaptableCache = Collections.synchronizedMap(new WeakHashMap<Class<?>, SoftReference<Object>>());
-                                adapterCache.put(cacheKey, adaptableCache);
-                            }
+                            Map<Class<?>, SoftReference<Object>> adaptableCache = getOrCreateCache(adaptable);
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:446]: Don't we already know here that there was no cache created for this `adaptable`? If a cache was available, we would have returned the result already - see https://github.com/apache/sling-org-apache-sling-models-impl/pull/21/files#diff-0c273224a9fa6e263f4381c5d4b8145aa2c6c89e75e0b62025eea5efa17edec9R401.

File: src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java (hunk 1 above)
Comment (cjelger) [reply src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:433]: No there could be a cache with the same `adaptable` but a different "target" model class. For example, one could first `adapt` the request to `MyModel.class` and then to `MyOtherModel.class`. So in the first case, the cache doesn't exist and the model is of course not cached, but in the 2nd case, the cache exists but the model is also not cached.

Comment (cjelger) [reply src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:433]: No, the same adaptable can be adapted to multiple types. For example, the request could be adapted first to `MyFirstModel.class` and then to `MySecondModel.class`: so in the first case there is no cache and of course no cached model, but in the 2nd case, the cache exists but doesn't contain a cached model for the `MySecondModel.class` (but would contain a cached model for `MyFirstModel.class`).

Comment (raducotescu) [reply src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java:433]: I think my question was not clear enough. I'm saying that if the `modelAnnotation#cache` returns `true`, we can already create the cache for the `adaptable` at https://github.com/apache/sling-org-apache-sling-models-impl/pull/21/files#diff-0c273224a9fa6e263f4381c5d4b8145aa2c6c89e75e0b62025eea5efa17edec9R401, using `getOrCreate`.

Further down, you just need to populate the cache. I think the `getCache` and `getOrCreate` should be merged and make the code simpler to read.

**Comment by cjelger** [comment]
@cziegeler @raducotescu I included your PR feedback: keep one single method to get or create the cache, and `synchronized` it on the adaptable to make sure we don't get any race condition. However SunarQube complains about the `synchronized` block because it's done on a method parameter, although here we exactly want to synchronize that block **_per_** adaptable. Not sure if this is a strict check ad I should fix or not. WDYT?

**Comment by raducotescu** [comment]
SonarCube is right and you have the rule's explanation: "The story is very similar for synchronizing on parameters; two different threads running the method in parallel could pass two different object instances in to the method as parameters, completely undermining the synchronization."

Usually you should synchronize on objects you control. For the case where the `adaptable` is a request you don't need to do any synchronization. For the other case I think it's sufficient to use `adapterCache.computeIfAbsent`.

**Comment by cjelger** [comment]
@raducotescu Just fixed the last issue, thx for your feedback.

**Pull request merged**